File: src/main/java/me/kikugie/ucsm/DirectionArgumentType.java
package me.kikugie.ucsm;

import com.mojang.brigadier.StringReader;
import com.mojang.brigadier.arguments.ArgumentType;
import com.mojang.brigadier.context.CommandContext;
import com.mojang.brigadier.exceptions.CommandSyntaxException;
import com.mojang.brigadier.exceptions.DynamicCommandExceptionType;
import com.mojang.brigadier.suggestion.Suggestions;
import com.mojang.brigadier.suggestion.SuggestionsBuilder;
import net.minecraft.command.CommandSource;
import net.minecraft.text.Text;
import net.minecraft.util.math.Direction;

import java.util.Collection;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.stream.Stream;

public class DirectionArgumentType implements ArgumentType<String> {
    public static final List<String> dirStrings = Stream.of(
            Direction.NORTH,
            Direction.WEST,
            Direction.SOUTH,
            Direction.EAST
    ).map(Direction::asString).toList();

    private final DynamicCommandExceptionType UNEXPECTED_VALUE_EXCEPTION = new DynamicCommandExceptionType(
            component -> Text.translatable("Invalid direction: %s", component)
    );

    @Override
    public String parse(StringReader reader) throws CommandSyntaxException {
        String remainder = reader.getRemaining().trim().toLowerCase();
        if (dirStrings.contains(remainder)) {
            return reader.readString();
        }
        throw UNEXPECTED_VALUE_EXCEPTION.create(remainder);
    }

    @Override
    public <S> CompletableFuture<Suggestions> listSuggestions(CommandContext<S> context, SuggestionsBuilder builder) {
        return CommandSource.suggestMatching(dirStrings, builder);
    }

    @Override
    public Collection<String> getExamples() {
        return dirStrings;
    }
}

File: src/main/kotlin/dev/kikugie/ucsm/command/DirectionArgumentType.kt
package dev.kikugie.ucsm.command

import com.mojang.brigadier.StringReader
import com.mojang.brigadier.arguments.ArgumentType
import com.mojang.brigadier.context.CommandContext
import com.mojang.brigadier.exceptions.CommandSyntaxException
import com.mojang.brigadier.exceptions.DynamicCommandExceptionType
import com.mojang.brigadier.suggestion.Suggestions
import com.mojang.brigadier.suggestion.SuggestionsBuilder
import net.minecraft.command.CommandSource
import net.minecraft.text.Text
import net.minecraft.util.math.Direction
import java.util.concurrent.CompletableFuture

class DirectionArgumentType : ArgumentType<Direction> {
    private val UNEXPECTED_VALUE_EXCEPTION =
        DynamicCommandExceptionType { component: Any -> Text.of("Invalid direction: $component") }

    @Throws(CommandSyntaxException::class)
    override fun parse(reader: StringReader): Direction {
        val remainder = reader.readUnquotedString()
        return if (remainder in dirStrings) Direction.byName(remainder)!! else
            throw UNEXPECTED_VALUE_EXCEPTION.create(remainder)
    }

    override fun <S> listSuggestions(
        context: CommandContext<S>,
        builder: SuggestionsBuilder
    ): CompletableFuture<Suggestions> {
        return CommandSource.suggestMatching(dirStrings, builder)
    }

    override fun getExamples() = dirStrings

    companion object {
        val dirStrings = Direction.entries.filter { it.axis.isHorizontal }.map { it.getName() }
    }
}
